src/cpp/core/Database.cpp

```diff
@@ -563,6 +563,42 @@ std::string PooledConnection::driverName() const
    return connection_->driverName();
 }
 
+ConnectionPool::ConnectionPool(const ConnectionOptions& options) :
+   connectionOptions_(options)
+{
+}
+
+void ConnectionPool::testAndReconnect(boost::shared_ptr<Connection>& connection)
+{
+   // do not test Sqlite connections - there is no backend system to connect to in this case
+   // so any errors on the file handle itself we do not want to gracefully recover from, as they would
+   // indicate a very serious programming error
+   if (connection->driver() == Driver::Sqlite)
+      return;
+
+   // it is possible for connections to go stale (such as if the upstream connection is closed)
+   // which will prevent it from being usable - we test for this by running a very efficient query
+   // and checking to make sure that no error has occurred
+   Error error = connection->executeStr("SELECT 1");
+   if (!error)
```

Comment:
Two questions here...
1) Do we want to preserve/log the error which caused us to attempt a reconnect?
2) Do we always want to reconnect regardless of the error returned? (i.e. are there some errors which are unrecoverable?)

Reply:
@kfeinauer Is there any reason why we can't use that list of recoverable Postgres errors that RSC uses here?
Another thing, could we preserve the last error that happened in the database connection instead of having to execute a new query?

Reply:
We need to execute this query whenever a connection is taken from the pool to make sure it is still alive. Unfortunately, the error codes that are returned by SOCI do not properly indicate if we had a connection issue - SOCI only returns that an `Unknown` error occurred on the connection. I verified this while testing - my original implementation depended on checking for that particular error state and re-opening the connection just for that error.

However, since `SELECT 1` should always succeed on a valid connection, this appears to be the most efficient way to determine that a connection error has actually occurred.

@jmcphers I will go ahead and log the original `SELECT 1` error - this will clutter the logs a little bit, but this should be a fairly rare occurrence and give us more diagnostic info should we need it.

Reply:
@kfeinauer It recall it was possible to "unwrap" the original Postgres error code. I think I used this:
https://github.com/SOCI/soci/blob/d4d50d65002ec2fb1486622634c31234b5f05686/src/backends/postgresql/error.cpp#L52
Look a little above that. You'll see that SOCI itself does some error classification. It don't quite match the one used in RSC though.

The reason why it's important to decide between recoverable and unrecoverable errors is that we shouldn't insist in reconnecting if the error is permanent, the service should stop. Insisting could throw the entire instance(s) in an undefined behavior state (functional, not in the C++ sense) potentially causing other problems or misleading users/admins. @jmcphers I'll let you make the call whether this is important for 1.4.

+1 for logging the `SELECT 1` error. But if we stop on unrecoverable errors, there won't be that many logs.

Reply:
I don't agree that we need to specifically look at the postgres error code here, and it would be more code to propagate that.

If you cannot successfully execute `SELECT 1`, then it's a connection issue. I think checking each and every error code like is done in the Connect case is a bit overkill here: https://github.com/rstudio/connect/blob/87ca7e50260fd48755101d4ed94dc509c22b965a/src/connect/store/store.go#L646. Certainly, encountering a scenario where there is a truly unrecoverable database error is very exceptional. 

Also, running sessions do not need database access to function properly - it's only needed to sign new users into the system. Stopping RSP entirely because of a database failure seems overkill to me, considering the vast majority of our product doesn't care if there's a database or not.

Reply:
I also do not know from looking at the sample code and the `postgresql_soci_error` code specifically how these error codes map and should exactly be handled. That introduces more uncertainty into the code than I am comfortable with at this point in the release.